### PR TITLE
chainload.py: transparently compress payload to improve loading time

### PR DESCRIPTION
**Pull request by svenpeter42**
Signed-off-by: Sven Peter <sven@svenpeter.dev>

### CODE REVIEW

File: proxyclient/chainload.py [hunk 1]
```diff
@@ -31,6 +32,23 @@
     # Fall back to proxy heap, which will be at the right place in old versions
     new_base = u.memalign(0x10000, memory_size)
 
+def compressed_writemem(iface, dest, data, progress):
```
Comment (marcan) [inline proxyclient/chainload.py:35]: This sounds generically useful, can we move it to utils and make it not leak memory? :-)

File: proxyclient/chainload.py [hunk 2]
```diff
@@ -31,6 +32,23 @@
...
+    if not len(data):
+        return
+
+    payload = lzma.compress(data)
+    compressed_size = len(payload)
+
+    try:
+        compressed_addr = p.malloc(compressed_size)
```
Comment (marcan) [inline proxyclient/chainload.py:43]: Just always use u.malloc; p.malloc is only useful when the data needs to outlive the execution of the script. I expect all python-side allocations to use the python-side allocator by default (it's faster too).

**Review by marcan — CHANGES_REQUESTED** [review]

File: proxyclient/chainload.py (hunk 1 above)
Comment (svenpeter42) [reply proxyclient/chainload.py]: Good idea! fixed.

File: proxyclient/chainload.py (hunk 2 above)
Comment (svenpeter42) [reply proxyclient/chainload.py]: I just blindly copied the code from above without understanding what was going on ;)
Fixed as well.

File: proxyclient/utils.py [hunk 3]
```diff
@@ -94,6 +94,21 @@ def inst(self, op):
...
+        payload = lzma.compress(data)
+        compressed_size = len(payload)
+        compressed_addr = self.malloc(compressed_size)
+
+        self.iface.writemem(compressed_addr, payload, progress)
+        decompressed_size = self.proxy.xzdec(compressed_addr, compressed_size, dest, len(data))
+
+        assert decompressed_size == len(data)
+
+        self.free(compressed_addr)
```
Comment (marcan) [inline proxyclient/utils.py:110]: Maybe try/finally guard this? Remember that malloc/free are Python-side, so e.g. if a proxy command fails and breaks m1n1 free will still work.

Comment (svenpeter42) [reply proxyclient/utils.py]: good idea, done.

Comment (marcan) [reply proxyclient/utils.py]: Nice, I like it!

**Review by marcan — APPROVED** [review]

**Pull request merged**